File: results/RQ1/05_baseline_qualification/units/p01_snake_game/python/replays/0001_seed_3736547328/repository/snake_game/snake.py
"""Mutable snake implementation for the snake game."""

from __future__ import annotations

from typing import List

from snake_game.direction import Direction
from snake_game.grid import Position


class Snake:
    """A mutable snake with a body and direction."""

    def __init__(self, start: Position, direction: Direction) -> None:
        """Initialize a snake with a starting position and direction.

        Args:
            start: The initial position of the snake's head.
            direction: The initial direction of the snake.
        """
        self._body: List[Position] = [start]
        self._direction = direction

    def turn(self, requested: Direction) -> bool:
        """Attempt to change the snake's direction.

        Args:
            requested: The requested direction to turn.

        Returns:
            True if the turn was successful (not reversing), False otherwise.
        """
        if self._direction.is_opposite(requested):
            return False
        self._direction = requested
        return True

    def head(self) -> Position:
        """Get the current head position of the snake.

        Returns:
            The current head position.
        """
        return self._body[0]

    def next_head(self) -> Position:
        """Calculate the next head position based on current direction.

        Returns:
            The position that would be the next head position.
        """
        current_head = self.head()
        dx, dy = self._direction.delta()
        return Position(current_head.x + dx, current_head.y + dy)

    def advance(self, grow: bool) -> None:
        """Advance the snake by moving its head and optionally growing its body.

        Args:
            grow: If True, the snake grows by one segment; if False, it moves without growing.
        """
        next_head = self.next_head()
        self._body.insert(0, next_head)
        if not grow:
            self._body.pop()

    def body(self) -> List[Position]:
        """Get the current body of the snake.

        Returns:
            A list of positions representing the snake's body.
        """
        return self._body[:]

File: results/RQ1/05_baseline_qualification/units/p01_snake_game/java/attempts/0001_seed_1332723230/p01_snake_game_java/p01_snake_game_java/src/main/java/Snake.java
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.List;

/**
 * Represents the snake in the game, maintaining its body as an ordered sequence of positions.
 * The body is stored as a Deque where the head is at the front and the tail at the back.
 */
public final class Snake {
    /** The body of the snake, represented as a deque of positions from head to tail. */
    private final Deque<Position> body;

    /**
     * Constructs a new snake with a single segment at the specified starting position.
     *
     * @param start the initial position of the snake's head
     * @throws NullPointerException if start is null
     */
    public Snake(Position start) {
        if (start == null) {
            throw new NullPointerException("Start position cannot be null");
        }
        this.body = new ArrayDeque<>();
        this.body.offerLast(start);
    }

    /**
     * Returns the current head position of the snake.
     *
     * @return the head position
     * @throws IllegalStateException if the snake has no body segments
     */
    public Position head() {
        if (body.isEmpty()) {
            throw new IllegalStateException("Cannot get head of empty snake");
        }
        return body.peekFirst();
    }

    /**
     * Returns the current tail position of the snake.
     *
     * @return the tail position
     * @throws IllegalStateException if the snake has no body segments
     */
    public Position tail() {
        if (body.isEmpty()) {
            throw new IllegalStateException("Cannot get tail of empty snake");
        }
        return body.peekLast();
    }

    /**
     * Returns the length of the snake in segments.
     *
     * @return the number of segments in the snake
     */
    public int length() {
        return body.size();
    }

    /**
     * Checks if the snake occupies the specified position.
     *
     * @param p the position to check
     * @return true if any segment of the snake occupies the position, false otherwise
     * @throws NullPointerException if p is null
     */
    public boolean occupies(Position p) {
        if (p == null) {
            throw new NullPointerException("Position cannot be null");
        }
        return body.contains(p);
    }

    /**
     * Moves the snake by adding a new head position and removing the tail.
     * This simulates the snake moving one cell in its current direction.
     *
     * @param nextHead the new head position
     * @throws NullPointerException if nextHead is null
     */
    public void moveTo(Position nextHead) {
        if (nextHead == null) {
            throw new NullPointerException("Next head position cannot be null");
        }
        body.offerFirst(nextHead);
        body.pollLast();
    }

    /**
     * Grows the snake by adding a new head position while keeping the tail.
     * This simulates the snake growing when it consumes food.
     *
     * @param nextHead the new head position
     * @throws NullPointerException if nextHead is null
     */
    public void growTo(Position nextHead) {
        if (nextHead == null) {
            throw new NullPointerException("Next head position cannot be null");
        }
        body.offerFirst(nextHead);
    }

    /**
     * Returns a defensive copy of the snake's body as a list, ordered from head to tail.
     *
     * @return a new list containing the snake's body positions in order
     */
    public List<Position> bodyAsList() {
        return new ArrayList<>(body);
    }
}